core/session.ts

```diff
@@ -42,9 +54,105 @@ export class Session {
     return adapters.create(this.config);
   }
 
+  public sqlxAction(
+    sqlxConfig: ISqlxConfig,
+    sqlStatementCount: number,
+    hasPreOperations: boolean,
+    hasPostOperations: boolean
+  ) {
+    if (sqlStatementCount > 1 && sqlxConfig.type !== "operations") {
+      const message = `Actions may only contain more than one SQL statement if they are of type 'operations'.`;
+      this.compileError(new Error(message));
+    }
+    if (sqlxConfig.hasOutput && sqlxConfig.type !== "operations") {
+      const message = `Actions may only specify 'hasOutput: true' if they are of type 'operations'.`;
+      this.compileError(new Error(message));
+    }
+    if (sqlxConfig.hasOutput && sqlStatementCount !== 1) {
+      const message = `Operations with 'hasOutput: true' must contain exactly one SQL statement.`;
+      this.compileError(new Error(message));
+    }
+    if (sqlxConfig.protected && sqlxConfig.type !== "incremental") {
+      const message = `Actions may only specify 'protected: true' if they are of type 'incremental'.`;
+      this.compileError(new Error(message));
+    }
+    if (sqlxConfig.disabled && !this.isDatasetType(sqlxConfig.type)) {
+      const message = `Actions may only specify 'disabled: true' if they create a dataset.`;
+      this.compileError(new Error(message));
+    }
+    if (sqlxConfig.redshift && !this.isDatasetType(sqlxConfig.type)) {
+      const message = `Actions may only specify 'redshift: { ... }' if they create a dataset.`;
+      this.compileError(new Error(message));
+    }
+    if (sqlxConfig.bigquery && !this.isDatasetType(sqlxConfig.type)) {
+      const message = `Actions may only specify 'bigquery: { ... }' if they create a dataset.`;
+      this.compileError(new Error(message));
+    }
+    if (hasPreOperations && !this.isDatasetType(sqlxConfig.type)) {
+      const message = `Actions may only include pre_operations if they create a dataset.`;
+      this.compileError(new Error(message));
+    }
+    if (hasPostOperations && !this.isDatasetType(sqlxConfig.type)) {
+      const message = `Actions may only include post_operations if they create a dataset.`;
+      this.compileError(new Error(message));
+    }
+
+    const action = (() => {
+      switch (sqlxConfig.type) {
+        case "view":
+        case "table":
+        case "inline":
+        case "incremental": {
+          const dataset = this.publish(sqlxConfig.name);
+          dataset.type(sqlxConfig.type);
```

Comment:
You could just call `dataset.config(sqlxConfig)` - that's the idea anyway! Then you remove some duplication here (this method is mostly the same as the config method in dataset.

Reply:
It would probably be nice to write a config({}) style method for operations and assertions too. This should make it easier to add or remove properties in the future for different types as this method just becomes routing to the correct action type.